dialog/dialog/relational-algebra.md

```diff
@@ -0,0 +1,415 @@
+    (category_id: 1, category_name: "Electronics", product_id: 2, product_category_id: 1, product_name: "Xerox 8"),
+    (category_id: 2, category_name: "Books", product_id: 3, product_category_id: 2, product_name: "Foucault's Pendulum"),
+    (category_id: 2, category_name: "Books", product_id: 4, product_category_id: 2, product_name: "100 Years of Solitude")
+]
+```
+
+## 3.10 Semijoin
+
+A semijoin is a binary operator which returns all tuples in its first input relation such that there is a tuple in its second relation for which the common attributes of both tuples are equal.
+
+This operation is equivalent to composing a [projection](#31-projection) again the first relation's attributes with a [natural join](#37-natural-join), however implementations are RECOMMENDED to implement semijoins as a specialization of these operations, for performance reasons.
+
+For example:
+
+```
+users = [
+    (user_id: 1, user_name: "Alice"),
+    (user_id: 2, user_name: "Bob"),
+    (user_id: 3, user_name: "Mallory")
+]
+
+admins = [
+    (user_id: 1),
+    (user_id: 2),
+]
+
+semijoin(users, admins) [
+    (user_id: 1, user_name: "Alice"),
+    (user_id: 2, user_name: "Bob"),
+]
+```
+
+## 3.11 Antijoin
+
+An antijoin is a binary operator which returns all tuples in its first input relation such that there is no tuple in its second relation for which the common attributes of both tuples are equal.
+
+This operation is equivalent to taking the [difference](#35-difference) of the first relation and the [semijoin](#310-semijoin) of both, however implementations MAY implement antijoins as a specialization of these operations, for performance reasons.
+
+## 3.12 Group By
+
+Group by is the mechanism by which aggregation is performed. Group by is performed over a relation with respect to a set of grouping attributes, and an aggregate function. The operation returns the set of tuples computed by first grouping the input relation, and then applying the aggregate function to each group.
+
+Implementations MAY support user defined aggregates, but MUST support the aggregate functions described in the [specification for the query language](query-engine.md#aggregation).
+
+For example:
+
+```
+products = [
+    (product_id: 1, product_category_id: 1, product_stock: 3),
+    (product_id: 2, product_category_id: 1, product_stock: 5),
+    (product_id: 3, product_category_id: 2, product_stock: 12),
+    (product_id: 4, product_category_id: 2, product_stock: 7)
+]
+
+# TODO Not a fan of this notation: returning a pair of the group and a scalar doesn't fit with the rest of the specification.
+#      In my prototype, aggregation is always over a variable, and the result is named using that variable. That's not
+#      how relational algebra is usually presented though, so I should revisit this as I describe more of what the overall
+#      compilation looks like.
+group_by(product_category_id, sum(product_stock), products) => [
```

Comment:
Not really a comment, more of a question:
I'm guessing
* you could have more than a single variable involved in the aggregation, e.g. `sum(product_stock * product_id)` (even though nonsensical), so naming the result as `product_stock` would be arbitrary/not really make sense
* the rest of the compilation pipeline doesn't make sense of multiple values of a tuple being involved in an aggregation?

Or is it perhaps just that your prototype didn't support multi-variable aggregation, since that wasn't needed for figuring out feasibility?

Reply:
I thought this might be the case, so this is a super useful comment that makes it clear to me that I'm not being clear here :)

There's a couple points to respond to here:

> you could have more than a single variable involved in the aggregation, e.g. sum(product_stock * product_id)

This is just a flub on my end, with how I documented RA. In Datalog, something like that would typically be written:

```
whatDoesThisEvenCompute(product_id: ProductId, result: Result) :-
    products(id: ProductId),
    Result = sum ProductStock * ProductId : products(id: ProductId, stock: ProductStock).
```

There, we explicitly bind the result of `sum` to a variable, and it's not implicit from the argument to `sum`.

Here's a similar example from Souffle:

```
B(z) :- z = min x+y : { A(x), A(y), C(y) }.
```

Supporting what you're asking about means being explicit, in RA, about which variable to bind the result to, and what expression to pass to the aggregate function, and that's absolutely something that needs to happen.

> Or is it perhaps just that your prototype didn't support multi-variable aggregation, since that wasn't needed for figuring out feasibility?

Nothing in my implementation precludes supporting aggregation over multiple variables in this position, like you're asking about, however I haven't implemented support for arithmetic operators, so the limitation is more to do with the fact that I don't have an expression language that can be used there.

There's a little bit of nuance that goes into guaranteeing a fixed point when leveraging an expression language like that; especially if the resulting tuples are being persisted anywhere. It's a solvable problem, but I don't have all the machinery for it in my prototype quite yet. Happy to chat with you more about that at some point!

Also: in case you're curious, in my implementation, I don't return a scalar from `group_by`, and instead return a tuple with a single attribute mapping the variable name to the aggregate result. That can then just be combined with the rest of the query using a `cartesian_product`!